src/develop/develop.h

```diff
@@ -150,6 +150,7 @@ typedef struct dt_develop_t
   // operations pipeline
   int32_t iop_instance;
   GList *iop;
+  dt_pthread_mutex_t fftw_lock;
```

Comment:
What do you need the mutex for? Is libfftw3 not thread safe?

Reply:
Hi, 
The most relevant part of libfftw, the plan execution, is thread safe. But the plan creation and plan destruction are not. Performance wise, this is not a problem, because the plan execution is the really significant portion. If you'd like to dive more into the details, you could have a look in [the doc](http://fftw.org/fftw3.pdf) on page 51, subsection 5.4. Cheers, Ingo

Reply:
And to add, I only use the mutex for plan creation at the very beginning and for plan destruction & cleanup at the end.

Reply:
I don't want to read the whole document now, but that section sounds as if the plan was something specific for the processed image at hand. Won't that break horribly when dt is processing two images in parallel, first creating the plan for image 1, releasing the lock, then creating the 2nd plan, then executing the plans. I expect both images working on the plan from the 2nd image. Or is the plan independent from the actual data and just some administrative data structures?

Reply:
It's the latter, only administrative data structures, depending exclusively on the tile size, which is constant. The plan is completely independent of the image content, so no worries in this respect. In theory it would even be possible to create a plan on darktable start-up and destroy it upon exit, but that would be at the expense of code readability in my opinion.

Reply:
creating "the plan" on darktable start-up would break parallel processing, wouldn't it?

Reply:
@houz : thanks for the hint, I'll look into it! Why I was not too enthusiastic about it: in demosaic.c, we define the tile size locally as `#define TS 122`. If you look at the plan creation, the only real dependency is on this tile size, as I said above. Now, if we move the plan creation to `init_global()`, this is not so ideal. If you have a nice suggestion for me, I'm more than open.

Reply:
@kubrickfr : I don't think we would break parallel processing. In the present structure, the plan creation is in `static void xtrans_fdc_interpolate`, but before the actual OpenMP parallelism for the tiles starts. So now there is one single plan creation per demosaic, and then parallel use of the one plan for the multiple tiles. So I see little risk here...

Reply:
OK, I think we can live with alternatives concerning the TS definition. One important question for @houz is: is single-threaded execution of `init_global()` and `cleanup_global()` guaranteed? Otherwise, we're still confronted with the need for a mutex.

Reply:
Yes, those are single threaded and called during dt initialization when loading the shared objects or closing dt.

Reply:
From my experience, that's the reason why it was crashing : when zooming in and out,  the demozaic function was being called before the previous call finished/was cancelled , which is what was causing the crashes. If you move plan memory allocation outside of the function, maybe you'll avoid the crashes but you'll get some corruption / glitches I bet.

Reply:
Seriously guys, you wanted us to break feature freeze and merge this code, despite that it apparently crashes during normal usage, *without* even mentioning that it  crashes during normal usage? As far as Ctrl+F can tell, this *is* the first mention.

Reply:
@LebedevRI, yes indeed I meant  "crashing without that current lock" :) sorry if my comment was misleading.
I have been using this patch for months now with thousands of pictures, frequently rebased against the master branch and it has been very stable.
The point I was trying to make is that I think moving memory allocation outside the function would break things. Anyway, what's so problematic with having a mutex here?

Reply:
So 
```diff
+// Preparations for fftw
+#ifdef HAVE_FFTW3
+  dt_pthread_mutex_lock(fftw_lock);
+  fftwf_complex *in_src = NULL, *out_src = NULL, *out_kernel = NULL, *in_kernel = NULL, *Cm = NULL;
+  // Initialization of the plans
+  fftwf_plan p_forw_src
+      = fftwf_plan_dft_2d(TS, TS, in_src, out_src, FFTW_FORWARD, FFTW_ESTIMATE); // real to complex
+  fftwf_plan p_forw_kernel = fftwf_plan_dft_2d(TS, TS, in_kernel, out_kernel, FFTW_FORWARD, FFTW_ESTIMATE);
+  fftwf_plan p_back = fftwf_plan_dft_2d(TS, TS, out_kernel, Cm, FFTW_BACKWARD, FFTW_ESTIMATE);
+  dt_pthread_mutex_unlock(fftw_lock);
+#endif
```
1. Does it actually depend on anything other than `TS` define?
2. Can you please replace that define with `static const int`.
3. Can the *exact same* plan be created once, and then used everywhere for everything, without creating it each time  in `xtrans_fdc_interpolate()`?

Reply:
@LebedevRI 3) as I mentioned before when zooming in and out (for example) interpolation functions such as xtrans_fdc_interpolate() get called before the previous cancelled call is finished. I do not believe this is specific to xtrans_fdc_interpolate()

Reply:
@kubrickfr that is orthogonal to my question

Reply:
@LebedevRI Indeed, therefore no, I haven't tried, but I don't think you can uses the exact same plan created once

Reply:
@LebedevRI : to answer your three questions:
1. No, the only dependency is on TS, as I already said above.
2. Yes, I can make such replacement if you prefer that, no problem. I took over the `#define` from the existing code above.
3. I am pretty sure that the exact same plan be created once, and then used everywhere for everything, without creating it each time in `xtrans_fdc_interpolate()`. Already now, it is created once and used multiple times for all the tiles, and one roi can have hundreds of tiles. Busy day for me today, but I'll make the modifications tomorrow or on Friday and let you know.

Reply:
I removed the mutex and placed the plan creation and plan destruction in `init_global()` and `cleanup_global()`, respectively, as suggested by @houz . It doesn't crash on my machine, and following FFTW's documentation, I don't expect it to. I will follow this u with a rebase, and cleanup, as suggested by @LebedevRI . As said, I always addressed all your review comments thus far, and if you find something else, just let me know.